[lldb][lldb-dap] Cleanup breakpoint filters.

oontvoo opened the pull request:
Details:
  - remove Swift breakpoint filter because this version of LLDB does not support Swift.
  - only return objc filters when working on macos.

jimingham commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1628,7 +1628,14 @@ void request_initialize(const llvm::json::Object &request) {
+  std::string triple =
+      std::string(g_dap.debugger.GetSelectedPlatform().GetTriple());
   for (const auto &exc_bp : g_dap.exception_breakpoints) {
+    // Skipping objc breakpoint filters if not working on macos.
```
GNUStep is still a thing.  It's likely the exception throw breakpoint is the only part of our ObjC support that would work with GNUStep, however, so maybe this doesn't matter all that much.

clayborg commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -36,9 +36,7 @@ DAP::DAP()
           {{"cpp_catch", "C++ Catch", lldb::eLanguageTypeC_plus_plus},
            {"cpp_throw", "C++ Throw", lldb::eLanguageTypeC_plus_plus},
            {"objc_catch", "Objective-C Catch", lldb::eLanguageTypeObjC},
-           {"objc_throw", "Objective-C Throw", lldb::eLanguageTypeObjC},
-           {"swift_catch", "Swift Catch", lldb::eLanguageTypeSwift},
-           {"swift_throw", "Swift Throw", lldb::eLanguageTypeSwift}}),
+           {"objc_throw", "Objective-C Throw", lldb::eLanguageTypeObjC}}),
```
See my comments for this patch in the review. It would be great to determine if a language is supported by LLDB dynamically and only add support for languages that the current LLDB supports. So we would probably need to default construct the `exception_breakpoints()` and then inside this constructor or before we try to use `exception_breakpoints`, we would need to populate it by doing something like:
```
if (g_dap.debugger.SupportsLanguage(lldb::eLanguageTypeC_plus_plus)) {
  exception_breakpoints.emplace_back({"cpp_catch", "C++ Catch", lldb::eLanguageTypeC_plus_plus})
  exception_breakpoints.emplace_back("cpp_throw", "C++ Throw", lldb::eLanguageTypeC_plus_plus})
}
if (g_dap.debugger.SupportsLanguage(lldb::eLanguageTypeObjC)) {
  exception_breakpoints.emplace_back({"objc_catch", "Objective-C Catch", lldb::eLanguageTypeObjC})
  exception_breakpoints.emplace_back("objc_throw", "Objective-C Throw", lldb::eLanguageTypeObjC})
}
if (g_dap.debugger.SupportsLanguage(lldb:: eLanguageTypeSwift)) {
  exception_breakpoints.emplace_back({"swift_catch", "Swift Catch", lldb:: eLanguageTypeSwift})
  exception_breakpoints.emplace_back("swift_throw", "Swift Throw", lldb:: eLanguageTypeSwift})
}
```

oontvoo replied:
Thanks for the detailed suggestions! Makes sense!

For ObjC, just to clarify, the intention of hiding the filters was NOT because the debugger didn't support it - the intention was to avoid clustering/confusing users - that is, if they're not debugging ObjC, there's no reason to show the objc filters. (And we infer this by checking if they're debugging on a mac - it's theoretically possible they're cross debugging objc from a non-mac but in practice I don't think it's likely )
Does this seem reasonable?

jimingham replied:
The ObjC Language runtime is able to detect whether ObjC is loaded into the program.  It runs that detector on each library load, so it keeps an accurate notion of this.  Ditto for the swift language runtime.  It would be better to consult that than guess based on platform.  There may not currently be a way to ask that from the SB API's but we can certainly make one...

oontvoo replied:
> The ObjC Language runtime is able to detect whether ObjC is loaded into the program. It runs that detector on each library load, so it keeps an accurate notion of this. Ditto for the swift language runtime. It would be better to consult that than guess based on platform.

This code runs before the runtimes (objc/swift) so it probably won't be able to query that, though?

oontvoo replied:
P.S: it was pointed out to me that changed capabilities can be pushed as an event. So I guess we could init the list of filters based on whether the languages are supported. Then when the runtime(s) become available, we can update the list after querying for which language is being debugged. WDYT?

jimingham replied:
ObjC tends to only use exceptions for error reporting, not for message passing, so a lot of ObjC developers leave this breakpoint on all the time.  Having to wait till the first load event that brings in ObjC would make that tedious.

walter-erquinigo replied:
Btw, it's possible to register typesystems/languages at initialization via runtime plugins. I think it would be ideal if the breakpoint list could get updated/refreshed after the target has been loaded (maybe at the first stop point)

clayborg reviewed (commented):
I would be nice if we can detect if we support Swift dynamically. Internally in LLDB, we can ask for a TypeSystem by language using:
```
llvm::Expected<lldb::TypeSystemSP>
TypeSystemMap::GetTypeSystemForLanguage(lldb::LanguageType language,
                                        Module *module, bool can_create);
llvm::Expected<lldb::TypeSystemSP>
TypeSystemMap::GetTypeSystemForLanguage(lldb::LanguageType language,
                                        Target *target, bool can_create);
```
These functions will return respectively:
```
TypeSystem::CreateInstance(language, module);
TypeSystem::CreateInstance(language, target);
```
We should be able to add a new static method to the TypeSystem.h/.cpp that can answer if a language is supported with something like:
```
static bool TypeSystem::SupportsLanguage(lldb::LanguageType language);
```
Each TypeSystem plugin can register a new callback that can answer if they support a language. Then we can add a static function on SBDebugger that would expose this function with something like:
```
static bool SBDebugger::SupportsLanguage(lldb::LanguageType language);
```
I will make comments in the DAP.cpp where this would be used.

adrian-prantl commented:
> I would be nice if we can detect if we support Swift dynamically. Internally in LLDB, we can ask for a TypeSystem by language using:

See also how lldb/test/Shell/Process/UnsupportedLanguage.test works

bulbazord commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -1628,7 +1628,14 @@ void request_initialize(const llvm::json::Object &request) {
   for (const auto &exc_bp : g_dap.exception_breakpoints) {
+    // Skipping objc breakpoint filters if not working on macos.
+    if (exc_bp.language == lldb::eLanguageTypeObjC &&
+        triple.find("macos") == std::string::npos) {
```
In addition to what Jim said above, this might not always work. For example, there are more apple triples (e.g. `arm64-apple-ios`)

oontvoo committed:
[lldb]Clean up breakpoint filters
 - added util function for querying whether a language is supported by the type system
 - populate the breakpoint filters table based on the supported language(s)

oontvoo commented:
> I would be nice if we can detect if we support Swift dynamically. Internally in LLDB, we can ask for a TypeSystem by language using:
> 
> ```
> llvm::Expected<lldb::TypeSystemSP>
> TypeSystemMap::GetTypeSystemForLanguage(lldb::LanguageType language,
>                                         Module *module, bool can_create);
> llvm::Expected<lldb::TypeSystemSP>
> TypeSystemMap::GetTypeSystemForLanguage(lldb::LanguageType language,
>                                         Target *target, bool can_create);
> ```
> 
> These functions will return respectively:
> 
> ```
> TypeSystem::CreateInstance(language, module);
> TypeSystem::CreateInstance(language, target);
> ```
> 
> We should be able to add a new static method to the TypeSystem.h/.cpp that can answer if a language is supported with something like:
> 
> ```
> static bool TypeSystem::SupportsLanguage(lldb::LanguageType language);
> ```
> 
> Each TypeSystem plugin can register a new callback that can answer if they support a language. Then we can add a static function on SBDebugger that would expose this function with something like:
> 
> ```
> static bool SBDebugger::SupportsLanguage(lldb::LanguageType language);
> ```
> 
> I will make comments in the DAP.cpp where this would be used.

Done!

oontvoo committed:
addressed review comments:
 - use std::optional
 - rename plugins->languages

JDevlieghere reviewed (approved):
LGTM!

oontvoo committed:
addressed review comment

oontvoo committed:
Merge branch 'llvm:main' into bp_filter

oontvoo commented:
@clayborg  Hi, do you have any further comments/feedback on this? Thanks!
(If not, I plan to merge this in the next few days)

gulfemsavrun commented:
We started seeing lldb test failures in `TestDAP*`.
```
======================================================================
FAIL: test_commands (TestDAP_launch.TestDAP_launch.test_commands)
    Tests the "initCommands", "preRunCommands", "stopCommands",
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/packages/Python/lldbsuite/test/decorators.py", line 150, in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/test/API/tools/lldb-dap/launch/TestDAP_launch.py", line 324, in test_commands
    self.continue_to_breakpoints(breakpoint_ids)
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py", line 245, in continue_to_breakpoints
    self.verify_breakpoint_hit(breakpoint_ids)
  File "/b/s/w/ir/x/w/llvm-llvm-project/lldb/packages/Python/lldbsuite/test/tools/lldb-dap/lldbdap_testcase.py", line 99, in verify_breakpoint_hit
    self.assertTrue(False, "breakpoint not hit")
AssertionError: False is not true : breakpoint not hit
```

felipepiovezan commented:
@oontvoo I think this broke the bots again: https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/5341/console

felipepiovezan commented:
Would you mind reverting it until you have a chance to look at the failures?
